### PR TITLE
Wildcard auditing

### PR DESCRIPTION
**Pull request by DL6ER**
**By submitting this pull request, I confirm the following (please check boxes, eg [X]) _Failure to fill the template will close your PR_:**

***Please submit all pull requests against the `development` branch. Failure to do so will delay or deny your request***

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [X] I have checked that [another pull request](https://github.com/pi-hole/FTL/pulls) for this purpose does not exist.
- [X] I have considered, and confirmed that this submission will be valuable to others.
- [X] I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
- [X] I give this submission freely, and claim no ownership to its content.

**How familiar are you with the codebase?:** 

## 10

---

Accept lines like `*.spot.com` as wildcard audits for domains.

One note about this:
- `*spot.com` will audit `spot.com`, `some.spot.com`, etc. but also `spotspot.com` or `totalldifferentspot.com`
- `*.spot.com` will audit `some.spot.com`, etc. but neither `spot.com` nor `spotspot.com` nor `totalldifferentspot.com`

_This template was created based on the work of [`udemy-dl`](https://github.com/nishad/udemy-dl/blob/master/LICENSE)._

### CODE REVIEW

File: grep.c [hunk 1]
```diff
@@ -189,9 +189,22 @@ int countlineswith(const char* str, const char* fname)
 	// Search through file
 	// getline reads a string from the specified file up to either a newline character or EOF
 	while(getline(&buffer, &size, fp) != -1)
-		if((strstr(buffer, str)) != NULL)
+	{
+		// Strip potential newline character at the end of line we just read
+		if(buffer[strlen(buffer)-1] == '\n')
+			buffer[strlen(buffer)-1] = '\0';
+
+		printf("Searching \"%s\" in \"%s\"\n",buffer,str);
```
Comment (Mcat12) [inline grep.c]: Was this for debugging? Also, the two variables might be switched.

File: grep.c [hunk 2]
```diff
@@ -189,9 +189,22 @@ int countlineswith(const char* str, const char* fname)
...
+		// Search for exact match
+		if(strcmp(buffer, str) == 0)
 			found++;
```
Comment (Mcat12) [inline grep.c]: If there was an exact match, we can probably `continue` here and skip the next `if`.

File: grep.c [hunk 3]
```diff
@@ -189,9 +189,22 @@ int countlineswith(const char* str, const char* fname)
...
 
+		// If line starts with *, search for partial match of
+		// needle "buffer+1" in haystack "str"
+		if(buffer[0] == '*' && strstr(str, buffer+1) != NULL)
```
Comment (Mcat12) [inline grep.c]: This will match more than just `xxxxxxexample.com` given `*example.com`. `example.com.xxxxx` will also match.

File: grep.c (hunk 1 above)
Comment (DL6ER) [reply grep.c]: It is `char *strstr(const char *haystack, const char *needle);`, so the variable ordering are correct.

**Comment by tuurie72** [comment]
After running the checkout to FTL (thanks DL6ER), I can confirm this is working as designed.
Love this feature and it saves a huge amount of time if you have a Sonos, Playstation, Spotify connection at your home!

**Pull request merged**